app/models/concerns/publishing_api/content.rb

```diff
@@ -62,17 +62,41 @@ def content
       values_from_json_paths = INDEXABLE_CONTENT_VALUES_JSON_PATHS.map do |item|
         item.on(document_hash).map { |body| BodyContent.new(body).html_content }
       end
+
       values_from_parts = document_hash.dig(:details, :parts)&.map do |part|
         # Add the part title as a heading to help the search model better understand the structure
         # of the content
         ["<h1>#{part[:title]}</h1>", BodyContent.new(part[:body]).html_content]
       end
 
-      [*values_from_json_paths, *values_from_parts]
+      [*values_from_json_paths, *values_from_parts, *values_from_blocks(document_hash)]
         .flatten
         .compact_blank
         .join(INDEXABLE_CONTENT_SEPARATOR)
         .truncate_bytes(INDEXABLE_CONTENT_MAX_BYTE_SIZE)
     end
+
+    def values_from_blocks(document_hash)
+      matches = JsonPath.new("$.details.blocks..content", use_symbols: true).on(document_hash)
+      return [] unless matches.any?
+
+      ignore_set = []
+      values = []
+      matches.each do |match|
+        case match
+        in Array
+          match.each { |m| ignore_set << m[:content] if m[:content].present? }
+          values << BodyContent.new(match).html_content
+        else
+          if ignore_set.index(match).present?
+            ignore_set.delete_at(ignore_set.index(match))
+          else
+            values << BodyContent.new(match).html_content
+          end
+        end
+      end
+
+      values
```

Comment:
You might be able to do this more elegantly using recursion instead of JsonPath....

Maybe something like this (currently untested):

https://github.com/alphagov/search-api-v2/pull/336/files

Reply:
Hmm, I think you'd want to explicitly match content_type: "text/html" in your code, otherwise you're potentially in the situation where a block with multiple options (text/html and text/govspeak) gets into the index twice. If you run the test suite, though, that should catch that situation.

Reply:
This code gets run after publishing-api has done it's `recursively_transform_govspeak` bit I assume? So it should always have `text/html` (and may or may not also have `text/govspeak`)?

Reply:
I like this @richardTowers – the problem @KludgeKML has been hitting (and me too on my brief attempts to dig into this) is the incomplete JSONPath support in the not actively maintained Ruby gem (insert classic XKCD here).

In theory, this single JSONPath expression should do the trick (find all `content` fields that either do not have a `content_type` subfield or have one and it's HTML):
```
$.details.blocks..content[?(!@.content_type || @.content_type == 'text/html')]
```

... and then we could have let the regular content extraction logic take over.

But there [isn't support for negation of subexpressions in the Ruby library](https://github.com/joshbuddy/jsonpath/issues/151) (there's a PR but it's not been merged in almost 2 years).

The JSONPath approach has been great for the bulk of static schema content (I still really like just having a giant list of possible paths!) but I begrudgingly accept we might not be able to make it work for blocks. If so the recursive approach you outlined seems neater than dealing with an ignore set (modulo some tidying to make it work).

Reply:
Gotcha, so I think my PR is fixed up / would work now. It's making pretty heavy use of [pattern matching](https://docs.ruby-lang.org/en/master/syntax/pattern_matching_rdoc.html), which might be a bit unfamiliar... It's part of ruby though, so I don't think we should be scared of it...

```ruby
def values_from_blocks_document(document_hash)
  if document_hash in { details: { blocks: [*blocks] } }
    blocks.flat_map { values_from_blocks(_1) }
  else
    []
  end
end

def values_from_blocks(input)
  case input
  in {blocks: [*blocks]}
    blocks.flat_map { values_from_blocks(_1) }
  in {content_type:, content:}
    content_type == "text/html" ? [content] : []
  in {content: String => content}
    [content]
  in Hash
    input.values.flat_map { values_from_blocks(_1) }
  in Array
    input.flat_map { values_from_blocks(_1) }
  else
    []
  end
end
```

Reply:
Actually doesn't need that second `in {blocks: [*blocks]}` bit either, as the Hash / Array fallbacks cover that.

Reply:
We already use pattern matching elsewhere in this app so we're all good. I think this can probably be tidied a little bit more, one sec ✍️

Reply:
How would you feel about:
```ruby
def values_from_blocks(item)
  case item
  in { content_type: "text/html", content: html_content }
    html_content
  in { content: String => content } unless item.key?(:content_type)
    content
  in Hash
    item.values.flat_map { values_from_blocks(_1) }
  in Array
    item.flat_map { values_from_blocks(_1) }
  else
    nil
  end
end
```

And then calling that at the bottom of `#content` like so:
```ruby
[
  *values_from_json_paths,
  *values_from_parts,
  *values_from_blocks(document_hash.dig(:details, :blocks)),
].flatten
  .compact_blank
  .join(INDEXABLE_CONTENT_SEPARATOR)
  .truncate_bytes(INDEXABLE_CONTENT_MAX_BYTE_SIZE)
```

Reply:
The question I was about to add was that what I needed to stop all the other tests failing was a guard clause checking for details/blocks outside of the recursive function, so that solution seems reasonable for me.

Reply:
@richardTowers Do you want to 
1) make your PR reviewable, then we can 
2)  merge it into this one, then I can 
3) update the description for this PR, and then 
4) Christian gives it the thumbs up and we're good to go?

Reply:
Sure, although I've got a bit of a run of meetings so you might have to wait a few hours. I'm totally fine for you to just fold the changes into this one and close my draft PR if that's easier.